Fixed #28529 -- Fixed VariableDoesNotExist warnings in admin templates.

ziima opened the pull request.

timgraham commented:
Did you consider how we might add tests for this to prevent regressions?

ziima commented:
I don't think that's possible at the moment. It would require a possibility to turn on strict mode for template rendering which would raise exceptions in case of exception occurs in template rendering.

Also I'm not entirely convinced it's necessary. The templates produce the same result regardless of their tidiness. Although it's better to have clean templates, I don't think it's so critical to add special tests for templates purity.

Possible improvement would be an option to turn on template errors when running tests of Django itself. I wasn't able to manage that and I had to add extra print statement to see them. That would definitely help to maintain clean templates.

ziima commented:
On another note: `is_popup` is the most commonly missing variable in admin contexts. I considered adding `is_popup=False` into `AdminSite.each_context`. Any thoughts on that?

timgraham commented:
"Possible improvement would be an option to turn on template errors when running tests of Django itself." That's what I meant by "add tests". Putting `is_popup=False` in `each_context()` seems okay.

ziima commented:
I modified the `each_context` and rebased.

The tests would probably be for another ticket, as I found no simple way to turn templates into strict mode.

jdufresne commented:
> as I found no simple way to turn templates into strict mode.

What about the technique described in the following mailing list thread:

https://groups.google.com/d/topic/django-developers/ZMzF0PgNb-k/discussion

> It as actually possible, just in a very hacky way. Django has a setting called TEMPLATE_STRING_IF_INVALID (or TEMPLATES[0]["OPTIONS"]["string_if_invalid"] in current versions of Django). If you override it and set it to a class with a __str__ that raises an exception you can get the effect you want. We use this in tests to find missing variables quickly.

ziima commented:
I saw that, but I consider it too big a hack to use it in tests. It's basically turning a `string_if_invalid` into a `raise_if_invalid` which is currently missing.

I'd rather leave it untested, until the strict template rendering is properly implemented.

jdufresne commented:
In template [`django/contrib/admin/templates/admin/edit_inline/tabular.html`](https://github.com/django/django/blob/6d1df84c00d45d382a95347c9d6cf25c40de572e/django/contrib/admin/templates/admin/edit_inline/tabular.html#L39) there is the variable:

```
{{ inline_admin_form.fk_field.field }}
```

[The `fk_field` method can sometimes return an empty string](https://github.com/django/django/blob/6d1df84c00d45d382a95347c9d6cf25c40de572e/django/contrib/admin/helpers.py#L347-L352). When it does, this template is referencing invalid attribute `field` on an empty string, which does not exist. Not sure if this counts as an variable error in the context of this PR, but I thought I would mention it. If so, you could wrap the variable rendering in a `{% if inline_admin_form.fk_field %}` statement.

jdufresne commented:
In the documentation [`docs/ref/templates/api.txt`](https://github.com/django/django/blob/f1c007bbf2fcd4996e29f0482c32faf5df397aa0/docs/ref/templates/api.txt#L380-L383), there is the blurb

> Many templates, including those in the Admin site, rely upon the silence of the template system when a nonexistent variable is encountered. If you assign a value other than ``''`` to ``string_if_invalid``, you will experience rendering problems with these templates and sites.

I think this should be updated as "including those in the Admin site" should now be inaccurate. Perhaps

> Many templates from third party applications rely upon the silence ...

ziima commented:
* `fk_field` treated in both `edit_inline` templates. Thanks for pointing it out.
* Regarding the docs, I would claim this just yet. Admin templates shouldn't anymore rely on `string_if_invalid`, but currently there's no way to be sure.

jdufresne commented:
> Admin templates shouldn't anymore rely on string_if_invalid, but currently there's no way to be sure.

But the phrase "including those in the Admin site" suggests to the reader that the admin app is intentionally taking advantage of the loose variable handling, which is no longer true.

jdufresne commented:
I ran the test suite with a very rough, very WIP strict template mode. A number of tests are still failing due to undefined variables. Only some are in the admin. Is the goals to clean up all undefined variables in the test suite, or just ones on the admin ones?

You can see my WIP branch at:

https://github.com/jdufresne/django/tree/templates-if-invalid

Here is the list of failed tests in strict mode.

[test-failure-strict-mode.txt](https://github.com/django/django/files/1385629/test-failure-strict-mode.txt)

ziima commented:
This ticket only tries to mitigate problems in `admin` application. Thus I've changed the docs to "including some of the Django's templates" which is true.

I'm looking forward to see the `if_invalid` in the upstream.

ziima commented:
Any progress?

r3m0t commented on django/contrib/admin/templates/admin/submit_line.html:
```diff
@@ -1,7 +1,7 @@
 <div class="submit-row">
 {% if show_save %}<input type="submit" value="{% trans 'Save' %}" class="default" name="_save" />{% endif %}
-{% if show_delete_link %}
+{% if show_delete_link and original %}
```
It looks like this change is unnecessary - if `show_delete_link` is `True` then `original` will always be set

ziima replied:
That doesn't seem to be true. When I removed the `original` I got several warnings from original being `None`.

r3m0t commented:
@ziima we often have weird hacks in tests to ensure that certain code paths are never reached, and the hacky value of `string_is_invalid` is a possible example.

ziima commented:
I got back to this ticket, but all I could do is just to rebase it.

I tried to hack the `string_if_invalid`, but failed hard. I couldn't found a case which wouldn't go though `FilterExpression.resolve` with `ignore_failures=True` so `None` is returned directly. Long story short, the `string_if_invalid` is ignored.
